internal/app/handler/dataplane/blobstorage/azure/azure.go

```diff
@@ -27,73 +38,100 @@ type Config struct {
-	blobClient, err := storage.NewBasicClient(config.BlobAccountName, config.BlobAccountKey)
-	if err != nil {
-		return nil, fmt.Errorf("error creating storage blobClient: %+v", err)
-	}
-	blob := blobClient.GetBlobService()
+func NewBlobStorage(config *Config, inputBlobPrefix, outputBlobPrefix string, eventMeta *documentstorage.EventMeta, env *module.Environment) (*BlobStorage, error) {
 	asb := &BlobStorage{
-		blobClient:       blob,
 		containerName:    config.ContainerName,
 		outputBlobPrefix: outputBlobPrefix,
 		inputBlobPrefix:  inputBlobPrefix,
 		eventMeta:        eventMeta,
+		accountName:      config.BlobAccountName,
+		accountKey:       config.BlobAccountKey,
+		env:              env,
 	}
 	return asb, nil
 }
 
 //PutBlobs puts a file into Azure Blob Storage
 func (a *BlobStorage) PutBlobs(filePaths []string) (map[string]string, error) {
-	container, err := a.createContainerIfNotExist()
-	if err != nil {
-		return nil, err
-	}
-
-	readStorageOptions := storage.BlobSASOptions{
-		BlobServiceSASPermissions: storage.BlobServiceSASPermissions{
-			Read: true,
-		},
-		SASOptions: storage.SASOptions{
-			Start:  time.Now().Add(time.Duration(-1) * time.Hour),
-			Expiry: time.Now().Add(time.Duration(24) * time.Hour),
-		},
-	}
-
 	blobSASURIs := make(map[string]string)
 
 	for _, filePath := range filePaths {
-		_, nakedFilePath := path.Split(filePath)
-		blobPath := helpers.JoinBlobPath(a.outputBlobPrefix, nakedFilePath)
+		filePathOutOfEnv := strings.Replace(filePath, a.env.OutputBlobDirPath, "", 1)
+		if filePathOutOfEnv[0] == '/' {
+			filePathOutOfEnv = filePathOutOfEnv[1:]
+		}
+		blobPath := helpers.JoinBlobPath(a.outputBlobPrefix, filePathOutOfEnv)
 		file, err := os.Open(filePath)
 		if err != nil {
 			return nil, fmt.Errorf("failed to read data from file '%s', error: '%+v'", filePath, err)
 		}
 		defer file.Close() // nolint: errcheck
-		blobRef := container.GetBlobReference(blobPath)
-		_, err = blobRef.DeleteIfExists(&storage.DeleteBlobOptions{})
+
+		stat, err := file.Stat()
 		if err != nil {
 			return nil, err
 		}
-		err = blobRef.CreateBlockBlobFromReader(file, &storage.PutBlobOptions{})
+
+		b := stat.Size()
+		kb := float64(b) / 1024
+		mb := float64(kb / 1024)
+		var timeout time.Duration
+		if mb > 5 {
+			timeout = time.Duration(mb) * 60 * time.Second
+		}
+		c := azblob.NewSharedKeyCredential(a.accountName, a.accountKey)
+		p := azblob.NewPipeline(c, azblob.PipelineOptions{
+			Retry: azblob.RetryOptions{
+				Policy:     azblob.RetryPolicyExponential,
+				MaxTries:   3,
+				TryTimeout: timeout,
+			},
+		})
+		URL, _ := url.Parse(
+			fmt.Sprintf("https://%s.blob.core.windows.net/%s", a.accountName, a.containerName))
+		containerURL := azblob.NewContainerURL(*URL, p)
+		ctx := context.Background()
+		_, err = containerURL.Create(ctx, azblob.Metadata{}, azblob.PublicAccessNone)
```

Comment:
Can we move the container create outside the `for` loop as the container used for the files will always be the same or have I misunderstood?

Reply:
I will take another look tomorrow. Basically, I moved inside the loop because the containerURL and thus the blockBlobUrl are all tied to a "pipeline" which defines the policies for the call (i.e. retries, timeouts, etc.) and I want to dynamically set the timeout based on the file size as it defaults to 60 seconds. The actual calls didn't seem to respect any additional timeout I put on the context, only the TryTimeout defined when creating the pipeline. There's definetly some optimisation I can do here, even if it's just to move the credentials outside the loop. Thanks

Reply:
The method `WithPipeline` did the trick, allowed me to change the pipeline per request so can move that block of code out :)